### ISSUE TITLE
State handling crash between card browser / note editor / template editor

### ISSUE THREAD

**Issue by mikehardy** [ankidroid/Anki-Android#5011]
I've seen this show up via ACRA, and the google play console release robot army caught it with video this time so I could figure it out.

###### Reproduction Steps

1. Open the card browser (it's empty in their tests, and they play with searches sometimes but the search was cleared)
2. Click + to add a card, enter text in front and back, template should be Basic
3. click the button bar that says "Card 1" to open the template editor
4. Click the options menu and select add to add a Card to the template, and you can mess with the template or not, but don't save - preview or not but don't save - then click the back button and agree it's okay to lose your progress
5. The Card button bar should say "Cards: Card 1 Card 2" now, but you never saved the second template
6. Click save to add the card, then click back to go back to the browser and it should crash with a JSONException string out of bounds

###### Expected Result

If you don't save the template the unsaved work should go away, you should have only Card 1 and the browser should work fine

###### Actual Result

Crash. I can reproduce this locally.

https://play.google.com/apps/publish/?account=5338425030304417978#PreLaunchReportPlace:p=com.ichi2.anki&appid=4973711737547064258&plrtab=CRASH&plrvc=20900140

FATAL EXCEPTION: AsyncTask #4
Process: com.ichi2.anki, PID: 12740
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$3.done(AsyncTask.java:353)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.RuntimeException: org.json.JSONException: Index 1 out of range [0..1)
	at com.ichi2.libanki.Card.template(Card.java:342)
	at com.ichi2.libanki.Card._getQA(Card.java:290)
	at com.ichi2.anki.CardBrowser.updateSearchItemQA(CardBrowser.java:1164)
	at com.ichi2.async.DeckTask.doInBackgroundSearchCards(DeckTask.java:818)
	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:262)
	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:69)
	at android.os.AsyncTask$2.call(AsyncTask.java:333)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 4 more
Caused by: org.json.JSONException: Index 1 out of range [0..1)
	at org.json.JSONArray.get(JSONArray.java:293)
	at org.json.JSONArray.getJSONObject(JSONArray.java:521)
	at com.ichi2.libanki.Card.template(Card.java:337)
	... 11 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at java.util.ArrayList.get(ArrayList.java:437)
	at org.json.JSONArray.get(JSONArray.java:287)
	... 13 more

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [ x ] character to confirm the points below:*

[  ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[  ] I have searched for similar existing issues here and on the user forum

**Comment by mikehardy** [comment]
Oh no. After you do this your database is corrupt. You can't open the CardBrowser again post-crash, it just keeps crashing. That explains why I've seen it in ACRA - even if rare it'll be over-represented because without AcraLimiter widely distributed this crash will show up until a user runs "Check Database".

Confirmed this affects 2.8.x as well.

@timrae I can fix the issue, but what's the policy for a database corruption bug on older versions? Do I hook into the version upgrade process and for all versions before the fixed one I hook check database, or (if I can swing it) a more targeted sql query/fix?

**Comment by mikehardy** [comment]
also - I'm thinking best course is to fix this on 2.8.5 then forward port, rather than fix in master and cherry-pick but I'm open to suggestion.

**Comment by mikehardy** [comment]
@timrae I haven't figured out why I'm not generating a clean PR to hotfix-2.8.5 yet, but here's the commits that fix the problem on 2.8.5 you can tell me if I've gone crazy while I go to a parent's school meeting, and I'll try a PR after

Integrity check:
https://github.com/mikehardy/Anki-Android/commit/6df95e989e87c0524eed1377a87aa229e67088bb

Integrity check logging:
https://github.com/mikehardy/Anki-Android/commit/363ab2631434c0f35389883cdb6468b7247336be

The bug fix:
https://github.com/mikehardy/Anki-Android/commit/1fc168f731e47a1c5ceb648444d9c9ef1ab2749a

**Comment by eginhard** [comment]
I just played around with this in the beta. It correctly reverts the model only if you click the back icon in the toolbar. When using the phone's back button it doesn't ask about losing my progress and persists the non-saved changes. You'll see some crashes from me because if you then delete the "Card 2" again and return using the phone's back button, it will think nothing has changed, not reload the browser and still show both cards, but interacting with the (actually deleted) second card causes the crash.

On the other hand, if you delete "Card 2" and then abandon that edit correctly (confirming that you want to lose progress), all second cards associated with this note type still remain deleted. You confirmed that when deleting "Card 2", but those deletes should be reverted as well.

AnkiDroid Version = 2.8.5beta1
Android Version = 6.0.1
ACRA UUID = 59d02946-b534-4858-a266-438ebc92d44b

**Comment by mikehardy** [comment]
2 thoughts - 1 - Great testing, thank you!, 2- nooooo :-)

**Comment by mikehardy** [comment]
I see your crashes which is useful for two reasons - @timrae  I made large changes to ACRA and they are being recorded along with other people's crashes - but only 1 crash (per crash type x per device), and I have the stack traces:

java.lang.RuntimeException: No card with id 1539160971068
at com.ichi2.libanki.Card.load(Card.java:146)
at com.ichi2.libanki.Card.<init>(Card.java:121)
at com.ichi2.libanki.Collection.getCard(Collection.java:454)
at com.ichi2.anki.CardBrowser$7.onItemClick(CardBrowser.java:492)
at android.widget.AdapterView.performItemClick(AdapterView.java:310)
at android.widget.AbsListView.performItemClick(AbsListView.java:1145)
at android.widget.AbsListView$PerformClick.run(AbsListView.java:3066)
at android.widget.AbsListView$3.run(AbsListView.java:3903)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:148)
at android.app.ActivityThread.main(ActivityThread.java:5461)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

**Comment by mikehardy** [comment]
We weren't handling the hardware back button, good catch. 

Also, if you came to the NoteEditor highlighted on Card3 (for example), then went into the CardTemplateEditor and deleted that one, the NoteEditor would blow up trying to underline Card3 in the card template button bar.

This thing is a bit of a nest of bugs

**Comment by eginhard** [comment]
Thanks, that PR looks good. That will still leave the bug I described in the second paragraph: when deleting a card type, all associated cards are deleted. When this change is then abandoned in the template editor, the card type is restored but its cards remain deleted.

**Comment by mikehardy** [comment]
Yeah - that's the last bit I'm working on now - pretty messy. I'm patching it up this style (which I will call "template editor persists speculative edits") but what a mess! At some point this thing needs to change to "persists when save is tapped"...

**Comment by timrae** [comment]
The workflow of submitting PR's directly to the hotfix branch has been bothering me for sometime, but I didn't say anything because I thought I was being pedantic. After seeing the evolution of this PR it's become crystal clear that my intuition was correct from the start. Please always submit your PR's to master branch. I'll cherry pick across to hotfix as I see fit when preparing the releases, after the code has been tested in alpha channel. The only exception where PR direct to hotfix would be necessary is non-cherry-pickable commits.

**Comment by timrae** [comment]
Oops that was intended to go in the PR not the issue

**Comment by mikehardy** [comment]
Hoping to get the related PR in shape soon - we're still crashing in this area in current alpha (2.9alpha47) from the pre-launch report.

The sequence the robot hit appeared to be:
- add note of basic type to empty deck
- while editing note go to edit cards
- add template in card template editor
- hit back button and agree to lose template edits
- note editor still shows the cards, hit save
- robot clicked around a bit but then clicked on default deck for review
- crash

```
10-19 14:50:26.476: E/AnkiDroid(17831): DeckTask/ doInBackgroundAnswerCard - RuntimeException on answering card
10-19 14:50:26.476: E/AnkiDroid(17831): java.lang.RuntimeException: org.json.JSONException: Index 1 out of range [0..1)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.libanki.Card.template(Card.java:342)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.libanki.Card._getQA(Card.java:290)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.libanki.Card._getQA(Card.java:282)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.async.DeckTask.doInBackgroundAnswerCard(DeckTask.java:501)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:247)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:69)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
10-19 14:50:26.476: E/AnkiDroid(17831): 	at java.lang.Thread.run(Thread.java:764)
```

**Comment by mikehardy** [comment]
Pushing these through in chunks, but in case I get hit by a bus I have code working to the point where there is no functionality regression between 2.8 and my PR, though I still don't handle undo in my branch.

https://github.com/mikehardy/Anki-Android/tree/issue5011-4

That tree is the base from which I'm peeling off bits and feeding them in PR by PR

**Comment by maryammsd** [comment]
I want this bug for my current research but i cannot reproduce it by downloading the versions of 2.8.2, 2.8.3 and 2.8.4. Is there any specific Android version this bug occurs at? Can you help me with this?

**Comment by mikehardy** [comment]
@maryammsd it was not merged until the 2.12 alphas series - 2.12alpha5 to be precise https://github.com/ankidroid/Anki-Android/commit/24a5a324683998f19138a5ed400d53e5341b8a66 - so any version prior to that should crash. The error was not device specific, so I must presume your reproduction scenario is either not sufficient to trigger it, or the corruption detection  you are using is not correctly determining whether something is corrupt or not
